base/server/python/pki/server/__init__.py

```diff
@@ -247,6 +247,52 @@ def restart(self):
         logger.debug('Command: %s', ' '.join(cmd))
         subprocess.check_call(cmd)
 
+    def run(self, as_current_user=False):
+
+        java_home = self.config['JAVA_HOME']
+        java_opts = self.config['JAVA_OPTS']
+
+        classpath = [
+            Tomcat.SHARE_DIR + '/bin/bootstrap.jar',
+            Tomcat.SHARE_DIR + '/bin/tomcat-juli.jar',
+            '/usr/lib/java/commons-daemon.jar'
+        ]
+
+        cmd = []
+
+        # by default run PKI server as systemd user
+        if not as_current_user:
```

Comment:
I'm not sure the context of this or why this is important. Why do we need to switch to the `systemd` user? What if we're running as an arbitrary user?

Perhaps we want to rewrite this option as `--user <username>` and make switching users something you explicitly do? (Without context...) It doesn't make sense to me to default to switching to another user and only not-switch on command.

Reply:
It's a bit complicated. Please take a look at these pages:

* https://www.dogtagpki.org/wiki/PKI_ACME_Container
* https://blog.openshift.com/jupyter-on-openshift-part-6-running-as-an-assigned-user-id/

Normally we want to run PKI server (i.e. Tomcat) with the same UID used by the systemd service since the server files are owned by that user. So, by default this script will switch to that user with a `sudo` command.

However, in OpenShift all applications will run with UID=&lt;random&gt; and GID=0 regardless of what we specify in the Dockerfile. The &lt;random&gt; user is only known at runtime, so we cannot specify it with `--user <random>` in the Dockerfile. Here we use `--as-current-user` to prevent the script from executing the `sudo` command (because it probably will fail anyway) so it will run the server as the current (i.e. &lt;random&gt;) user. Perhaps we can pick a less confusing parameter name?

FYI, to let the server access the server files, we execute this command in the Dockerfile:

```
RUN chgrp -Rf root /var/lib/tomcats/acme && chmod -Rf g+w /var/lib/tomcats/acme
```
The above command gives the root group (i.e. GID=0) permission to access the server files.

Reply:
"with the same UID used by the systemd service" -- are you saying, `pkiuser`? I'm thinking that a containers are a special case; I'd reasonably expect (in a container, without `systemd`) `pkispawn` to create things as the current user, so by default you'd want to run as the current user.

If you want to switch to a particular user (e.g., `pkispawn`) if it exists in the container and you want to install it that way, you could use something like `--user <user>`.

Perhaps I'm still thinking about this wrong or that there needs to be additional work to enable this to what I'd expect. I guess it is fine as-is, just thought it was really weird.

Reply:
Yes, for basic Tomcat server it will be `tomcat` user, and for full PKI server it will be `pkiuser`. Let's call this PKI_USER for simplicity.

Containers are indeed a special case, so it should be OK to require additional parameters (i.e. `pki-server run --as-current-user`) in the Dockerfile. The `pki-server run` command will probably more often be used by admins who want to troubleshoot issues by running the server in the foreground (e.g. with jdb). Since the server files are owned by PKI_USER, and we want the server to be able to access the files, and we don't want the command to generate new files (e.g. log files) that cannot be accessed by PKI_USER later, we should run the server as PKI_USER by default.

Requiring the admin to specify `pki-server run --user PKI_USER` would be rather inconvenient since they need to figure out the value and make sure it is correct since mistakes could create permission errors somewhere that is not easy to find.

Running `pkispawn --user <OpenShift user>` is a completely different issue though. `pkispawn` has to be invoked from the Dockerfile since it can only be run by root. At that point the actual OpenShift user is not yet known, so we cannot specify that in the Dockerfile. `pkispawn` will have to use the normal PKI_USER, then we use the trick described in [this page](https://blog.openshift.com/jupyter-on-openshift-part-6-running-as-an-assigned-user-id/) to give the &lt;OpenShift user&gt; permissions to access the server files.